# Module 06 Lab - Regression and Classification Models**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. ClassificationThis is the most fundamental distinction between types of supervised learning problems.*   **Regression:** The goal is to predict a **continuous numerical value**.     *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.*   **Classification:** The goal is to predict a **discrete category or class label**.    *   *Examples:* Predicting if an email is spam or not spam, if a flower is a setosa, versicolor, or virginica, or if a customer will churn or not.**In this lab, we will tackle one of each.**

## Part 2: Linear Regression**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Load and prepare the data
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

# For simplicity, we'll drop rows with missing age
df.dropna(subset=['Age'], inplace=True)

# Define features and target
features = ['Age', 'Pclass']
target_reg = 'Fare'
X_reg = df[features]
y_reg = df[target_reg]

# Split the data
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

### Task 1: Train and Evaluate a Linear Regression Model**Your Task:**1.  Create an instance of the `LinearRegression` model.2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).3.  Make predictions on the test data.4.  Evaluate the model using `mean_squared_error`. This metric tells us the average of the squared differences between the predicted and actual values.

In [ ]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Create the Linear Regression model.
lr_model = LinearRegression()

# Train the model so it can learn the relationship between age, class, and fare.
lr_model.fit(X_train_reg, y_train_reg);

# Predict fares for the passengers in the test set.
y_pred_reg = lr_model.predict(X_test_reg)

# Evaluate the size of the prediction errors.
mse = mean_squared_error(y_test_reg, y_pred_reg)
rmse = np.sqrt(mse)

print(f"Mean Squared Error for Fare Prediction: {mse:.2f}")
print(f"Root Mean Squared Error: ${rmse:.2f}")
print(f"Model coefficients [Age, Pclass]: {lr_model.coef_}")
print(f"Model intercept: {lr_model.intercept_:.2f}")

Mean Squared Error for Fare Prediction: 3364.92
Root Mean Squared Error: $58.01
Model coefficients [Age, Pclass]: [ -0.43695889 -38.29728935]
Model intercept: 133.31


## Part 3: Logistic Regression**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It works by calculating the probability that a given input belongs to a certain class. It's one of the most widely used and interpretable classification models.*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Convert the text categories in Sex into numeric values for the model.
# male = 0 and female = 1
df["Sex"] = df["Sex"].map({"male": 0, "female": 1})

# Select the input features and categorical target for classification.
features_cls = ["Age", "Pclass", "Sex"]
target_cls = "Survived"

X_cls = df[features_cls]
y_cls = df[target_cls]

# Split the classification data into training and testing sets.
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(
    X_cls,
    y_cls,
    test_size=0.20,
    random_state=42
)

print(f"Classification training rows: {len(X_train_cls)}")
print(f"Classification testing rows: {len(X_test_cls)}")


Classification training rows: 571
Classification testing rows: 143


### Task 2: Train and Evaluate a Logistic Regression Model**Your Task:**1.  Create an instance of the `LogisticRegression` model.2.  Train the model using the classification training data.3.  Make predictions on the test data.4.  Evaluate the model using `accuracy_score`.

In [ ]:
# Create the Logistic Regression classifier.
# A larger max_iter value gives the training process enough iterations to converge.
log_model = LogisticRegression(max_iter=1000)

# Train the classifier using the passenger characteristics and survival labels.
log_model.fit(X_train_cls, y_train_cls)

# Predict whether each passenger in the test set survived.
y_pred_cls = log_model.predict(X_test_cls)

# Calculate the percentage of test predictions that were correct.
accuracy = accuracy_score(y_test_cls, y_pred_cls)

print(f"Accuracy for Survival Prediction: {accuracy:.2%}")

Accuracy for Survival Prediction: 74.83%


## 📝 Knowledge Check**Instructions:** Answer the following questions in this markdown cell.
1.  **In your own words, what is the key difference between a regression problem and a classification problem?** A regression problem predicts a continuous numerical value, while a classification problem predicts a category or class. In this lab, predicting a passenger's fare is regression because the answer can be many different dollar amounts. Predicting survival is classification because the result is one of two labels: survived or did not survive.

2.  **The `LinearRegression` model has an attribute called `.coef_`. After you train the model, print `lr_model.coef_`. What do these numbers represent?** The coefficients show how much the model's predicted fare changes when one feature increases by one unit while the other feature stays the same. The Age coefficient is approximately 0.44, which means the predicted fare decreases by about $0.44 for each additional year of age when passenger class is held constant. The Pclass coefficient is approximately -38.30, which means moving from one class number to the next, such as first class to second class, lowers the predicted fare by about 38.30 when age is held constant. These values describe the model's learned relationships and do not prove that either feature directly causes a fare change.

3.  **Why did we use `mean_squared_error` to evaluate the regression model but `accuracy_score` for the classification model?**  Mean squared error is appropriate for regression because it measures the numerical distance between each predicted fare and the actual fare. Squaring the errors also gives larger mistakes more weight. Accuracy is appropriate for classification because it measures the percentage of correctly predicted class labels. Accuracy would not work well for fare prediction because fares are continuous values, and a prediction such as 49.50 would normally not exactly match an actual fare of $50.00 even though the prediction is very close.